<a href="https://colab.research.google.com/github/Deli8t/data/blob/master/Delight_Assignmet_Introduction_to_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Notebook -Introduction to Regression Project



# Background Information
**Problem Statement**

Mobile carrier Megaline has found out that many of their subscribers use legacy plans.
They want to develop a model that would analyze subscribers' behavior and recommend
one of Megaline's newer plans: Smart or Ultra.
You have access to behavior data about subscribers who have already switched to the
new plans (from the project for the Statistical Data Analysis course). For this
classification task, you need to develop a model that will pick the right plan. Since you’ve
already performed the data preprocessing step, you can move straight to creating the
model.
Develop a model with the highest possible accuracy. In this project, the threshold for
accuracy is 0.75. Check the accuracy using the test dataset.

1. Open and look through the data file.
2. Split the source data into a training set, a validation set, and a test set.
3. Investigate the quality of different models by changing hyperparameters. Briefly
describe the findings of the study.
4. Check the quality of the model using the test set.
5. Additional task: sanity check the model. This data is more complex than what
you’re used to working with, so it's not an easy task. We'll take a closer look at it
later.

**Dataset**

● Dataset URL (CSV File): https://bit.ly/UsersBehaviourTelco

# 1.Data Exploration

In [30]:
#We will now import  libraries we require  to read our csv file 
# from an external  source. 

import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier

user_behaviour = pd.read_csv("https://bit.ly/UsersBehaviourTelco")
                 
user_behaviour.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [ ]:
#Checking the type of a dataset
user_behaviour.dtypes

calls       float64
minutes     float64
messages    float64
mb_used     float64
is_ultra      int64
dtype: object

In [ ]:
#Checking the size of a dataset
user_behaviour.shape

(3214, 5)

In [ ]:
#checking for null values
user_behaviour.isna().sum()

calls       0
minutes     0
messages    0
mb_used     0
is_ultra    0
dtype: int64

In [ ]:
#information about the data frame
user_behaviour.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [ ]:
#checking for the duplicate values
any(user_behaviour.duplicated())

False

In [ ]:
#describing the dataset
user_behaviour.describe()

calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

In [ ]:
#checking for missing values
round(100*(user_behaviour.isnull().sum()/len(user_behaviour.index)), 2).sort_values(ascending = False) 

calls       0.0
minutes     0.0
messages    0.0
mb_used     0.0
is_ultra    0.0
dtype: float64

In [6]:
print(user_behaviour.describe())
print()
print('The number of non-ultra customers is:', user_behaviour.is_ultra.loc[user_behaviour.is_ultra == 0].count())
print()
print('The number of ultra customers is:', user_behaviour.is_ultra.loc[user_behaviour.is_ultra == 1].count())
print()
print('The percentage of non-ultra customers is:', (len(user_behaviour.is_ultra.loc[user_behaviour.is_ultra == 0])/len(user_behaviour.is_ultra)*100),'%')
print()
print('The percentage of ultra customers is:', (len(user_behaviour.is_ultra.loc[user_behaviour.is_ultra == 1])/len(user_behaviour.is_ultra)*100),'%')

             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246     0.461100
min       0.000000     0.000000     0.000000      0.000000     0.000000
25%      40.000000   274.575000     9.000000  12491.902500     0.000000
50%      62.000000   430.600000    30.000000  16943.235000     0.000000
75%      82.000000   571.927500    57.000000  21424.700000     1.000000
max     244.000000  1632.060000   224.000000  49745.730000     1.000000

The number of non-ultra customers is: 2229

The number of ultra customers is: 985

The percentage of non-ultra customers is: 69.35283136278781 %

The percentage of ultra customers is: 30.647168637212197 %


In [11]:
#Creating a target variable.
x_true = user_behaviour.is_ultra.values

#Printing how the variable itself + len of the variable.
print(x_true)
print()
print(len(x_true)) #Should return the len of the starting dataframe..is_ultra.values

#Printing how the variable itself + len of the variable.
print(x_true)
print()
print(len(x_true)) #Should return the len of the starting dataframe.

[0 0 0 ... 0 0 1]

3214
[0 0 0 ... 0 0 1]

3214


#2.Data Preparation

In [21]:
# Splitting the data into train and test.
user_behaviour_train, user_behaviour_test = train_test_split(user_behaviour, stratify=x_true, test_size=0.25, random_state=12345)

#Splitting the training dataset into train validation and cross validation.
user_behaviour_train, user_behaviour_valid = train_test_split(user_behaviour_train, test_size=0.25, random_state=12345)

In [22]:
#Train and Cross validation dataset are going to be used to build the model itself.
#Test is going to check the accuracy of our model output.

print('Number of records in training dataset:', user_behaviour_train.shape[0])
print('Number of records in cross validation dataset:', user_behaviour_valid.shape[0])
print('Number of records in the test dataset:', user_behaviour_test.shape[0])
print('')


Number of records in training dataset: 1807
Number of records in cross validation dataset: 603
Number of records in the test dataset: 804



In [23]:
#To make sure that our testing is going to have a positive result,
#We can save the column in a variable to compare that to the model results obtained.

checking_values = user_behaviour_test.is_ultra #This is going to become the array to check the accuracy of our model.

In [24]:
#Creating features and targets for our model.

features_train = user_behaviour_train.drop(['is_ultra'], axis=1)
target_train = user_behaviour_train['is_ultra']
features_valid = user_behaviour_valid.drop(['is_ultra'], axis=1)
target_valid = user_behaviour_valid['is_ultra']
features_test = user_behaviour_test.drop(['is_ultra'], axis=1)
target_test = user_behaviour_test['is_ultra']


In [25]:
#Let's take a look to the distribution of is_ultra column in the three dataset that we should make sure, that everyone of our datasets 
#has at least a 25% of ultra customers in it. 
#To change this split is just necessary changing random_state parameter in train_test_split function

train_is_ultra_distribution = user_behaviour_train['is_ultra'].value_counts(normalize=True)
cv_is_ultra_distribution = user_behaviour_valid['is_ultra'].value_counts(normalize=True)
test_is_ultra_distribution =  checking_values.value_counts(normalize=True)

print('Train dataset:')
print(train_is_ultra_distribution)
print()
print('Validation dataset:')
print(cv_is_ultra_distribution)
print()
print('Testing array:')
print(test_is_ultra_distribution)


Train dataset:
0    0.691201
1    0.308799
Name: is_ultra, dtype: float64

Validation dataset:
0    0.699834
1    0.300166
Name: is_ultra, dtype: float64

Testing array:
0    0.69403
1    0.30597
Name: is_ultra, dtype: float64


# 3.Data Modeling

In [36]:
#Investigating the quality of different models by changing hyperparameters. 
#We will use these models; DecisionTreeClassifier / RandomForestClassifier / LogisticRegression

#Finding accuracy score for DecisionTreeClassifier
result = []

for depth in range(1, 10):
        model = DecisionTreeClassifier(random_state = 8, max_depth=depth)

#Create a model, specify max_depth=depth >
        model.fit(features_train, target_train)
#Train the model >
        predictions_valid = model.predict(features_valid) 
#Find the predictions using validation set >        
        print("max_depth =", depth, ": ", end='')
        acc = accuracy_score(target_valid, predictions_valid)
        print(acc)
        result.append(acc)
print()

print('Accuracy Score :', max(result))



max_depth = 1 : 0.75787728026534
max_depth = 2 : 0.7910447761194029
max_depth = 3 : 0.8009950248756219
max_depth = 4 : 0.7960199004975125
max_depth = 5 : 0.7943615257048093
max_depth = 6 : 0.8126036484245439
max_depth = 7 : 0.8009950248756219
max_depth = 8 : 0.8092868988391376
max_depth = 9 : 0.8009950248756219

Accuracy Score : 0.8126036484245439


In [34]:
#Finding Rmse for RandomForestClassifier
model = RandomForestClassifier(n_estimators=10,max_depth=10, random_state=8)
#Train the model >
model.fit(features_train, target_train) 
#Predict
predictions_valid = model.predict(features_valid) 

acc_forest = accuracy_score(target_valid, predictions_valid)
rmse_forest = mean_squared_error(target_valid, predictions_valid)**0.5
print('RMSE :', rmse_forest) #the lowest, the best.
print('Accuracy score:', acc_forest)

RMSE : 0.4552986372567942
Accuracy score: 0.7927031509121062


In [37]:
#Finding Rmse values for LogisticRegression

model = LogisticRegression(random_state=8, solver='liblinear', max_iter=100)
#Train the model >
model.fit(features_train, target_train) 
#Predict
predictions_valid = model.predict(features_valid) 

acc_logistic = accuracy_score(target_valid, predictions_valid)
rmse_logistic = mean_squared_error(target_valid, predictions_valid)**0.5
print("RMSE:", rmse_logistic) 
print('Accuracy score:', acc_logistic)

RMSE: 0.5402536103680887
Accuracy score: 0.7081260364842454


In [40]:
#Checking  the quality of the model using the test set
#Testing the best model on the test dataset.

#RandomForest
model = RandomForestClassifier(n_estimators=10,max_depth=10, random_state=8)
#Train the model >
model.fit(features_train, target_train) # < train the model >
predictions_test = model.predict(features_test)

print(f'Test: {model.score(features_test, target_test):.3f}')
print(f'Train: {model.score(features_train, target_train):.3f}')
print()
print('This model, obtained an 79% accuracy on test dataset, while 89% on the trained dataset.')

Test: 0.792
Train: 0.891

This model, obtained an 79% accuracy on test dataset, while 89% on the trained dataset.


In [42]:
#LogisticRegression
model = LogisticRegression(random_state=8, solver='liblinear', max_iter=100)
#Train the model >
model.fit(features_train, target_train) 
#Predict
predictions_valid = model.predict(features_valid) 

print(f'Test: {model.score(features_test, target_test):.3f}')
print(f'Train: {model.score(features_train, target_train):.3f}')
print()
print('This model, obtained an 71% accuracy on test dataset, while 71% on the trained dataset.')

Test: 0.711
Train: 0.717

This model, obtained an 71% accuracy on test dataset, while 71% on the trained dataset.


In [45]:
#DecisionTreeClassifier
result = []

for depth in range(1, 10):
        model = DecisionTreeClassifier(random_state = 8, max_depth=depth)

#Create a model, specify max_depth=depth >
        model.fit(features_train, target_train)
#Train the model >
        predictions_valid = model.predict(features_valid) 

print(f'Test: {model.score(features_test, target_test):.3f}')
print(f'Train: {model.score(features_train, target_train):.3f}')
print()
print('This model, obtained an 79% accuracy on test dataset, while 87% on the trained dataset.')



Test: 0.792
Train: 0.879

This model, obtained an 79% accuracy on test dataset, while 87% on the trained dataset.


In [46]:
#Sanity check the model.
#Predictions array.
print(predictions_test)
print(len(predictions_test))

[0 0 1 0 0 0 0 0 1 0 1 0 1 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 1 0 1 0 1
 1 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 1 0
 0 0 0 1 0 0 1 1 1 0 1 0 0 0 0 0 1 0 0 0 1 1 0 0 1 0 1 1 0 1 0 0 0 0 0 0 0
 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 1 1 0 0 0 0 0 0 0 0 1 0
 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0 1 1 1 0 1 1 0 0 0
 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0 0 0 0 0 0 1 0 0 1 1 0 0 0 1 1 0 0 0 1 0 0 0
 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 0 1 0 0 0 1 0 1 0 0 1
 1 0 1 1 0 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 1 0 0 0 0 0 0 1 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 0 1 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 1 1 0 0
 0 0 0 0 1 0 0 0 0 0 1 0 

In [47]:
#Checking values array.
print(checking_values.values)
print(len(checking_values))

[0 1 1 1 0 0 0 0 1 1 0 0 1 0 0 1 0 1 0 1 1 0 0 0 0 0 1 0 1 1 0 0 1 0 1 1 0
 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 0 1 1 0 0 0 1 0 0 0 1 1 0 0 1 1 0 1 0 0 0 1
 0 0 0 0 0 1 0 1 1 0 0 1 0 0 0 0 1 0 0 0 1 1 1 0 1 0 1 1 0 1 0 0 1 1 0 0 0
 0 0 1 0 1 0 0 0 0 0 0 0 1 0 0 0 0 1 1 0 0 0 1 1 0 1 1 0 0 0 1 0 0 0 0 0 0
 0 1 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 0 1 1 0 1 1 1 0 0
 0 1 0 0 0 0 0 0 1 0 0 0 0 1 1 0 1 1 0 0 1 1 1 0 0 1 0 1 1 0 0 0 1 0 0 1 0
 0 0 0 1 0 0 0 1 0 0 0 0 0 0 1 0 0 0 1 0 1 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 1
 0 1 0 0 1 0 1 0 1 1 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 1 0 1 0 0 0 0 0 0 0 1 1
 0 0 0 0 0 0 1 0 1 1 1 0 0 0 0 0 0 0 1 0 1 0 0 0 0 0 1 0 0 1 1 0 0 1 0 0 0
 1 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 1 0 1 0 0 0 0 0 1 1 0 1
 1 1 1 0 1 0 0 1 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 1 0 1 0 0 0 0 0 1 0 0 0 0 0
 1 0 0 1 0 0 0 1 1 1 1 1 0 0 1 0 1 0 0 1 1 0 1 1 0 1 0 0 0 0 0 0 0 0 0 1 0
 0 0 0 1 1 0 0 0 0 0 1 1 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 0 0
 0 0 0 0 1 0 0 0 0 0 0 0 

In [48]:
score = accuracy_score(target_test, predictions_test)
print(score)

0.7922885572139303


# 4.Summary of Findings and Recommendations

I analyzed the dataset , trying to build a model with the maximum possible accuracy to predict the plan of a certain customer, whether ultra or non-ultra.

After choosing, training and testing the model(DecisionTreeClassifier) I was able to obtain 81% of accuracy in my model. 79% on Test data  and 87% on Train data. 
